### ISSUE TITLE
monotonic slow memory increase after upgraded to 1.91.3 from 1.87.6

### ISSUE THREAD

**Issue by gabber78** [VictoriaMetrics/VictoriaMetrics#4690]
### Describe the bug

This monotonic slow memory increase happened after we upgraded vmauth (among other VictoriaMetrics components) to 1.91.3 from 1.87.6

Dumped memory via the pprof endpoint [here](https://github.com/gabber78/vm-memory-dump/blob/f2319ae965ae20be4a529e3056a229ce70ea1bd1/mem.pprof)

### To Reproduce

In 2 days time self occur.

### Version

1.91.3

### Screenshots

Since the version upgrade:
![image](https://github.com/VictoriaMetrics/VictoriaMetrics/assets/7890124/7a534b9c-4bb4-4911-92b8-588c9d92e20a)

Lat 14 days:

![image](https://github.com/VictoriaMetrics/VictoriaMetrics/assets/7890124/2073b97e-bddd-47cd-b02d-50f342934fc6)

Seems in around 2 days time (this is second cycle for us) the pods getting OOMKilled when mem consumption reach 100%.

**Comment by f41gh7** [comment]
Hello, may I ask you to create 2 memory profiles:
1) after 10-15 mins of  vmauth start
2) after 3-4 days of vmauth running.


 Also, can you plot graph for vmauth resident memory usage:

`rate(process_resident_memory_bytes{job="VMAUTH_JOB_NAME"})`

**Comment by gabber78** [comment]
Hi, unfortunately I can only provide memory profiles after 10 + 15 mins of vmauth start. Pods getting OOMKILLED after approximately 48hrs. If its help we can create a memory profile when we start getting the alerts (after pods getting killed in few hours).

Memory profiles are [HERE](https://github.com/gabber78/vm-memory-dump)

**Comment by gabber78** [comment]
plot graph for vmauth resident memory usage:

rate(process_resident_memory_bytes{pod="vmalert-skr-77"}[5m]) / 1024 / 1024

![image](https://github.com/VictoriaMetrics/VictoriaMetrics/assets/7890124/e97c5e6c-ab86-43cc-ac87-97ac3cf90188)

**Comment by talszabi** [comment]
Please find below-attached memory profiles after some (~3-4) days of no memory issues. 

[vmauth_mem_pprof_20230803.tar.gz](https://github.com/VictoriaMetrics/VictoriaMetrics/files/12249260/vmauth_mem_pprof_20230803.tar.gz)

**Comment by lumi017** [comment]
Faced the same issue today again, after restart it's back to low memory usage

**Comment by f41gh7** [comment]
I've checked memory profiles and memory usage is really low. 

 I think high memory consumption may be related to the sidecar container. Could you plot graph for per container memory usage?

**Comment by berkanercan** [comment]
Today the same issue occured again. Here is the graph and the memory profile. 
<img width="1204" alt="Screenshot 2023-08-11 at 10 17 29" src="https://github.com/VictoriaMetrics/VictoriaMetrics/assets/18217868/c35decaf-e488-4e52-bc4f-996d1837674e">

[11.08.2023_vmauth_mem.pprof.zip](https://github.com/VictoriaMetrics/VictoriaMetrics/files/12319945/11.08.2023_vmauth_mem.pprof.zip)

**Comment by ebensom** [comment]
@f41gh7 do you have any news after checking the latest mem profile where the vmauth container uses 400+ MiB?

**Comment by f41gh7** [comment]
> @f41gh7 do you have any news after checking the latest mem profile where the vmauth container uses 400+ MiB?

 Sry, didn't have free time for that. Most memory spent on regexp objects. Probably there is some leak at regexp usage. I hope to return with some results at the next week.

**Comment by hagen1778** [comment]
@ebensom @berkanercan could you please share cmd-line flags used for vmauth? Are there any suspicious logs for vmauth?

**Comment by ebensom** [comment]
@hagen1778 sure, the command-line flags for vmauth are:

```
    - -auth.config=/opt/vmauth/config.yaml
    - -httpListenAddr=:8427
    - -logInvalidAuthTokens=true
```

I don't see any suspicious logs during the last few days when the memory was slowly increasing in all vmauth replicas. One thing to note is that we have a quite dynamic setup of using 1300+ VMRules to generate the configuration dynamically by vm-operator, and there are frequent changes to the vmauth configurations due to addition/removal of VMUsers (a few dozen times per hour), in these cases config-reloading is being triggered:

```
  | Sep 26, 2023 @ 13:32:01.062 | 2023-09-26T13:32:00.214Z	info	VictoriaMetrics/app/vmauth/auth_config.go:353	loaded information about 1311 users from -auth.config="/opt/vmauth/config.yaml"
  | Sep 26, 2023 @ 13:32:00.606 | 2023-09-26T13:31:59.610Z	info	VictoriaMetrics/app/vmauth/auth_config.go:353	loaded information about 1311 users from -auth.config="/opt/vmauth/config.yaml"
  | Sep 26, 2023 @ 13:32:00.289 | 2023-09-26T13:31:59.583Z	info	VictoriaMetrics/app/vmauth/auth_config.go:328	SIGHUP received; loading -auth.config="/opt/vmauth/config.yaml"
  | Sep 26, 2023 @ 13:32:00.289 | 2023-09-26T13:31:59.691Z	info	VictoriaMetrics/app/vmauth/auth_config.go:353	loaded information about 1311 users from -auth.config="/opt/vmauth/config.yaml"
  | Sep 26, 2023 @ 13:32:00.238 | 2023-09-26T13:31:59.511Z	info	VictoriaMetrics/app/vmauth/auth_config.go:328	SIGHUP received; loading -auth.config="/opt/vmauth/config.yaml"
  | Sep 26, 2023 @ 13:32:00.238 | 2023-09-26T13:31:59.884Z	info	VictoriaMetrics/app/vmauth/auth_config.go:353	loaded information about 1311 users from -auth.config="/opt/vmauth/config.yaml"
  | Sep 26, 2023 @ 13:32:00.071 | 2023-09-26T13:31:59.512Z	info	VictoriaMetrics/app/vmauth/auth_config.go:328	SIGHUP received; loading -auth.config="/opt/vmauth/config.yaml"
  | Sep 26, 2023 @ 13:32:00.071 | 2023-09-26T13:31:59.563Z	info	VictoriaMetrics/app/vmauth/auth_config.go:353	loaded information about 1311 users from -auth.config="/opt/vmauth/config.yaml"
  | Sep 26, 2023 @ 13:32:00.062 | 2023-09-26T13:31:59.558Z	info	VictoriaMetrics/app/vmauth/auth_config.go:328	SIGHUP received; loading -auth.config="/opt/vmauth/config.yaml"
  | Sep 26, 2023 @ 13:32:00.039 | 2023-09-26T13:31:59.532Z	info	VictoriaMetrics/app/vmauth/auth_config.go:328	SIGHUP received; loading -auth.config="/opt/vmauth/config.yaml"
  | Sep 26, 2023 @ 13:32:00.039 | 2023-09-26T13:31:59.657Z	info	VictoriaMetrics/app/vmauth/auth_config.go:353	loaded information about 1311 users from -auth.config="/opt/vmauth/config.yaml"
  | Sep 26, 2023 @ 13:31:59.886 | 2023-09-26T13:31:59.512Z	info	VictoriaMetrics/app/vmauth/auth_config.go:328	SIGHUP received; loading -auth.config="/opt/vmauth/config.yaml"
  | Sep 26, 2023 @ 13:31:59.886 | 2023-09-26T13:31:59.563Z	info	VictoriaMetrics/app/vmauth/auth_config.go:353	loaded information about 1311 users from -auth.config="/opt/vmauth/config.yaml"
  | Sep 26, 2023 @ 13:31:59.715 | 2023-09-26T13:31:59.558Z	info	VictoriaMetrics/app/vmauth/auth_config.go:353	loaded information about 1311 users from -auth.config="/opt/vmauth/config.yaml"
  | Sep 26, 2023 @ 13:31:59.715 | 2023-09-26T13:31:59.511Z	info	VictoriaMetrics/app/vmauth/auth_config.go:328	SIGHUP received; loading -auth.config="/opt/vmauth/config.yaml"

```

**Comment by ebensom** [comment]
However we had the frequent config change/reload when using the previous 1.87.6 version, but we didn't notice any memory leak with that, only when upgrading to 1.91.3 we started to notice the issue (our vmauth pods are OOM killed and restarted now every 2-3 days).
We will upgrade to 1.94.5 next week and see if the problem still persists.

**Comment by hagen1778** [comment]
Closing as stale

**Comment by ebensom** [comment]
Hi @hagen1778 ,

Yes, we are currently running on 1.96.0 and the issue is still there.
The leak appears to be in the vmauth config (re)loading code, when using the YAML unmarshal (decoder). It looks like the regexp compiler instances and onePassCopy compilers are still held with references and not garbage collected. Attached the SVG of the memory profile indicating the code path to the leaks. 

![pprof001](https://github.com/VictoriaMetrics/VictoriaMetrics/assets/11090449/26d5acfe-751c-4ec0-911c-74fa023d38bb)

I wanted to troubleshoot and fix this issue during the Christmas holidays, I just didn't find the time yet.

**Comment by valyala** [comment]
FYI, the bugfix, which should fix the memory leak at `vmauth` on config reload, has been merged into `master` branch - see 88329d84ca7417ea3fdef03fcdf55d2b18c489ff . This commit also fixes exposition of `vmauth_unauthorized_user_concurrent_requests_capacity`, `vmauth_unauthorized_user_concurrent_requests_current`, `vmauth_user_concurrent_requests_capacity` and `vmauth_user_concurrent_requests_current` [metrics](https://docs.victoriametrics.com/vmauth.html#monitoring) after [config reload](https://docs.victoriametrics.com/vmauth.html#config-reload). Previously these metrics weren't updated after config reload.

This commit will be included in the next release. In the mean time you can build `vmauth` from this commit according to [these docs](https://docs.victoriametrics.com/vmauth/#how-to-build-from-sources) and verify whether it stops leaking memory on config reloads.

**Comment by valyala** [comment]
The memory leak in `vmauth` mentioned above has been fixed in [v1.98.0](https://docs.victoriametrics.com/CHANGELOG.html). Closing the issue as fixed then.

@gabber78 , please re-open the issue if vmagent v1.98.0 still has this memory leak.

**Comment by valyala** [comment]
FYI, the memory leak at `vmauth` has been also fixed in [v1.97.2 LTS release](https://github.com/VictoriaMetrics/VictoriaMetrics/releases/tag/v1.97.2).
